# KNN-3

Importing the necessary libraries:

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
from sklearn.model_selection import train_test_split

Q1. Write a Python code to implement the KNN classifier algorithm on load_iris dataset in
sklearn.datasets.

In [10]:
from sklearn.datasets import load_iris

X,y = load_iris(return_X_y=True)
df=pd.DataFrame(X, columns=load_iris().feature_names)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [11]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=42)
X_train.shape, X_test.shape

((120, 4), (30, 4))

In [12]:
knn_classifier=KNeighborsClassifier()
knn_classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [15]:
y_pred=knn_classifier.predict(X_test)

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print(f'confusion Matrix: \n{confusion_matrix(y_pred=y_pred,y_true=y_test)}\n')
print(f'Accuracy score : {accuracy_score(y_pred,y_test).round(3)}\n')
print(classification_report(y_pred,y_test))

confusion Matrix: 
[[10  0  0]
 [ 0  9  0]
 [ 0  0 11]]

Accuracy score : 1.0

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        10
           1       1.00      1.00      1.00         9
           2       1.00      1.00      1.00        11

    accuracy                           1.00        30
   macro avg       1.00      1.00      1.00        30
weighted avg       1.00      1.00      1.00        30



Q2. Write a Python code to implement the KNN regressor algorithm on load_boston dataset in
sklearn.datasets.

Ans Solution using pyhton is as shown:

In [17]:
try:
    from sklearn.datasets import load_boston
except Exception as e:
    print(f'Error: {e}')

Error: 
`load_boston` has been removed from scikit-learn since version 1.2.

The Boston housing prices dataset has an ethical problem: as
investigated in [1], the authors of this dataset engineered a
non-invertible variable "B" assuming that racial self-segregation had a
positive impact on house prices [2]. Furthermore the goal of the
research that led to the creation of this dataset was to study the
impact of air quality but it did not give adequate demonstration of the
validity of this assumption.

The scikit-learn maintainers therefore strongly discourage the use of
this dataset unless the purpose of the code is to study and educate
about ethical issues in data science and machine learning.

In this special case, you can fetch the dataset from the original
source::

    import pandas as pd
    import numpy as np

    data_url = "http://lib.stat.cmu.edu/datasets/boston"
    raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
    data = np.hstack([raw_df.values[::2, :]

Since, the dataset is not directly available, importing using method described by Scikit Learn:

In [79]:
data_url = "http://lib.stat.cmu.edu/datasets/boston"
raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
target = raw_df.values[1::2, 2]
columns=['CRIM','ZN','INDUS','CHAS','NOX','RM','AGE','DIS','RAD','TAX','PTRATIO','B','LSTAT']
boston_df=pd.DataFrame(data, columns=columns)
target_df=pd.DataFrame(target,columns=['MEDV'])
boston_df=pd.concat([boston_df, target_df], axis=1)
boston_df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

Variables in order:

- CRIM:     per capita crime rate by town
- ZN:       proportion of residential land zoned for lots over 25,000 sq.ft.
- INDUS:    proportion of non-retail business acres per town
- CHAS:     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
- NOX:      nitric oxides concentration (parts per 10 million)
- RM:       average number of rooms per dwelling
- AGE:      proportion of owner-occupied units built prior to 1940
- DIS:      weighted distances to five Boston employment centres
- RAD:      index of accessibility to radial highways
- TAX:      full-value property-tax rate per \$10,000
- PTRATIO:  pupil-teacher ratio by town
- B:        1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
- LSTAT:    % lower status of the population
- MEDV:     Median value of owner-occupied homes in $1000's

In [80]:
boston_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


In [85]:
X=boston_df.iloc[:,:-1]
y=boston_df['MEDV']

In [86]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=42)
X_train.shape, X_test.shape

((404, 13), (102, 13))

In [87]:
knn_regressor=KNeighborsRegressor()
knn_regressor.fit(X_train,y_train)

KNeighborsRegressor()

In [88]:
y_pred=knn_regressor.predict(X_test)

from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
print(f'R2 Score : {r2_score(y_test, y_pred).round(3)}')
print(f'Mean Absolute Error : {mean_absolute_error(y_test, y_pred).round(3)}')
print(f'Mean Squared Error : {mean_squared_error(y_test, y_pred).round(3)}')

R2 Score : 0.647
Mean Absolute Error : 3.664
Mean Squared Error : 25.86


Q3. Write a Python code snippet to find the optimal value of K for the KNN classifier algorithm using cross-validation on load_iris dataset in sklearn.datasets.

In [70]:
from sklearn.datasets import load_iris

X,y = load_iris(return_X_y=True)
df=pd.DataFrame(X, columns=load_iris().feature_names)
df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

In [71]:
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.20, random_state=56)
X_train.shape, X_test.shape

((120, 4), (30, 4))

In [76]:
from sklearn.model_selection import GridSearchCV

param={'n_neighbors': list(range(1,12))}

clf=GridSearchCV(KNeighborsClassifier(), param_grid=param, verbose=3)
clf.fit(X_train, y_train)

Fitting 5 folds for each of 11 candidates, totalling 55 fits
[CV 1/5] END .....................n_neighbors=1;, score=0.917 total time=   0.0s
[CV 2/5] END .....................n_neighbors=1;, score=0.958 total time=   0.0s
[CV 3/5] END .....................n_neighbors=1;, score=0.958 total time=   0.0s
[CV 4/5] END .....................n_neighbors=1;, score=0.917 total time=   0.0s
[CV 5/5] END .....................n_neighbors=1;, score=1.000 total time=   0.0s
[CV 1/5] END .....................n_neighbors=2;, score=0.917 total time=   0.0s
[CV 2/5] END .....................n_neighbors=2;, score=0.958 total time=   0.0s
[CV 3/5] END .....................n_neighbors=2;, score=0.917 total time=   0.0s
[CV 4/5] END .....................n_neighbors=2;, score=0.917 total time=   0.0s
[CV 5/5] END .....................n_neighbors=2;, score=1.000 total time=   0.0s
[CV 1/5] END .....................n_neighbors=3;, score=0.917 total time=   0.0s
[CV 2/5] END .....................n_neighbors=3;

GridSearchCV(estimator=KNeighborsClassifier(),
             param_grid={'n_neighbors': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]},
             verbose=3)

In [77]:
clf.best_params_

{'n_neighbors': 7}

In [78]:
clf.best_score_

0.9833333333333334

Thus the best value for K is 7 from Grid Search CV and the best accuracy score is 0.9833

Q4. Implement the KNN regressor algorithm with feature scaling on load_boston dataset in sklearn.datasets.

In [89]:
boston_df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [97]:
X=boston_df.iloc[:,:-1]
y=boston_df['MEDV']

from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_scaled=scaler.fit_transform(X)

X_train, X_test, y_train, y_test=train_test_split(X,y, test_size=0.20, random_state=23)
X_train.shape, X_test.shape

((404, 13), (102, 13))

In [112]:
knn_reg=KNeighborsRegressor(n_neighbors=5, metric='euclidean')
knn_reg.fit(X_train, y_train)

y_pred=knn_reg.predict(X_test)

In [113]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
print(f'R2 Score : {r2_score(y_test, y_pred).round(3)}')
print(f'Mean Absolute Error : {mean_absolute_error(y_test, y_pred).round(3)}')
print(f'Mean Squared Error : {mean_squared_error(y_test, y_pred).round(3)}')

R2 Score : 0.522
Mean Absolute Error : 4.32
Mean Squared Error : 41.157


Q5. Write a Python code snippet to implement the KNN classifier algorithm with weighted voting on
load_iris dataset in sklearn.datasets.

In [114]:
# Load the Iris dataset
data = load_iris()
X = data.data  # Features
y = data.target  # Target labels

# Split the data into a training set and a test set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

# Standardize the features
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [115]:

# Initialize the KNN classifier with weighted voting (weights='distance')
knn_classifier = KNeighborsClassifier(n_neighbors=5, weights='distance')

# Fit the model to the training data
knn_classifier.fit(X_train, y_train)


KNeighborsClassifier(weights='distance')

In [116]:
# Predict the labels for the test set
y_pred = knn_classifier.predict(X_test)


print(f'confusion Matrix: \n{confusion_matrix(y_pred=y_pred,y_true=y_test)}\n')
print(f'Accuracy score : {accuracy_score(y_pred,y_test).round(3)}\n')
print(classification_report(y_pred,y_test))

confusion Matrix: 
[[19  0  0]
 [ 0 13  0]
 [ 0  0 13]]

Accuracy score : 1.0

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        19
           1       1.00      1.00      1.00        13
           2       1.00      1.00      1.00        13

    accuracy                           1.00        45
   macro avg       1.00      1.00      1.00        45
weighted avg       1.00      1.00      1.00        45



Q6. Implement a function to standardise the features before applying KNN classifier.

In [68]:
def standardize_features(X_train, X_test):
    from sklearn.preprocessing import StandardScaler
    
    # Initialize the StandardScaler
    scaler = StandardScaler()
    # Fit the scaler to the training data and transform it
    X_train_standardized = scaler.fit_transform(X_train)
    # Transform the test or validation data using the same scaler
    X_test_standardized = scaler.transform(X_test)

    return X_train_standardized, X_test_standardized


In [69]:
# Example usage:
import numpy as np

# Create sample training and test data
X_train = np.array([[1, 2], [3, 4], [5, 6]])
X_test = np.array([[7, 8]])

# Standardize the features
X_train_standardized, X_test_standardized = standardize_features(X_train, X_test)

print("Standardized Training Data:")
print(X_train_standardized)
print("\nStandardized Test Data:")
print(X_test_standardized)

Standardized Training Data:
[[-1.22474487 -1.22474487]
 [ 0.          0.        ]
 [ 1.22474487  1.22474487]]

Standardized Test Data:
[[2.44948974 2.44948974]]


Q7. Write a Python function to calculate the euclidean distance between two points.

Ans. Solution using python is as follows:

In [59]:
def euclidean(a,b):
    from numpy import sqrt
    d2=0
    for i in range(len(a)):
        d2=d2+(a[i]-b[i])**2
    d=sqrt(d2)
    return d.round(4)

Example of use:

In [61]:
distance=euclidean((2,3), (4,6))
print(f'Euclidean distance between points (2,3) and (4,6) is {distance}')

Euclidean distance between points (2,3) and (4,6) is 3.6056


Q8. Write a Python function to calculate the manhattan distance between two points.

In [57]:
def manhattan(a,b):
    d=0
    for i in range(len(a)):
        d=d + abs(a[i]-b[i])
    return d

In [58]:
distance=manhattan((2,3), (4,6))
print(f'Manhattan distance between points (2,3) and (4,6) is {distance}')

Manhattan distance between points (2,3) and (4,6) is 5
